### DISCUSSION TITLE
Log raw json-rpc messages

### DISCUSSION THREAD

**Discussion by chioni16** [ebkalderon/tower-lsp#394] [Q&A] [answered]
Hello! Is it possible to log the raw[ json-rpc ](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#message) of messages both received from the client and that are sent back to client? 
Thanks!

**Comment by ebkalderon** [comment, marked as answer]
Hey there! `tower-lsp` includes built-in integration with [`tracing`](https://github.com/tokio-rs/tracing) for logging. The raw JSON-RPC payloads of both incoming (client-to-server) and outgoing (server-to-client) messages are recorded as `trace!()` logs here:

https://github.com/ebkalderon/tower-lsp/blob/7b580128faa39b907db8f40bc53924181887acde/src/codec.rs#L141

and here:

https://github.com/ebkalderon/tower-lsp/blob/7b580128faa39b907db8f40bc53924181887acde/src/codec.rs#L181

respectively.

Depending on which `tracing` subscriber you have set up, you should be able to filter by the "trace" log level emitted from `tower_lsp::codec` to view the raw messages. For instance, the example programs included in the [`examples/`](https://github.com/ebkalderon/tower-lsp/tree/7b580128faa39b907db8f40bc53924181887acde/examples) directory are set up using the `fmt` subscriber. To view the raw JSON-RPC message traffic, launch an example with `RUST_LOG=error,tower_lsp::codec=trace` and those log messages should be printed to `stderr`.

Hope this is helpful!

**Reply by chioni16** [reply]
Thanks!

**Reply by ebkalderon** [reply]
No problem! Happy to help. :smile:
